### ISSUE TITLE
Maven repository issues

### ISSUE THREAD

**Issue by g4s8** [artipie/maven-adapter#113]
There're some issues with Maven repository: it doesn't have any validation, `<versioning>` of metadata file may be incorrect if someone uploaded artifact just with `PUT` request, metadata file can be broken if user uploads valid `xml` metadata file with invalid artifact id and group id.

Consider changing maven deploy workflow to:
 1. Upload artifacts and checksum to temporary location, until `maven-metadata.xml` uploaded (always the last file of maven deploy)
 2. Perform some validation: check checksums are correct, check metadata group and id is the same as in repo `maven-metadata.xml` file, check metadata versions are correct.
 3. Merge new `maven-metadata.xml` file into current repo `maven-metadata.xml`, generate checksums for new metadata and rewrite it in temporary location
 4. Move all uploaded items and merged metadata file with updated checksums from temporary location to repository location

**Comment by g4s8** [comment]
The common deploy requests HTTP flow:
```
PUT /repo/com/test/logger/1.1/logger-1.1.jar HTTP/1.1
PUT /repo/com/test/logger/1.1/logger-1.1.jar.sha1 HTTP/1.1
PUT /repo/com/test/logger/1.1/logger-1.1.jar.md5 HTTP/1.1
PUT /repo/com/test/logger/1.1/logger-1.1.pom HTTP/1.1
PUT /repo/com/test/logger/1.1/logger-1.1.pom.sha1 HTTP/1.1
PUT /repo/com/test/logger/1.1/logger-1.1.pom.md5 HTTP/1.1
GET /repo/com/test/logger/maven-metadata.xml HTTP/1.1
GET /repo/com/test/logger/maven-metadata.xml.sha1 HTTP/1.1
PUT /repo/com/test/logger/maven-metadata.xml HTTP/1.1
PUT /repo/com/test/logger/maven-metadata.xml.sha1 HTTP/1.1
PUT /repo/com/test/logger/maven-metadata.xml.md5 HTTP/1.1
```

**Comment by 0crat** [comment]
@0crat[/z](https://www.0crat.com/u/0crat) I see that you unassigned this issue; the order is still assigned to @olenagerasimova[/z](https://www.0crat.com/u/olenagerasimova) though; to cancel the order use `refuse`, as in [§6](http://www.zerocracy.com/policy.html#6)

**Comment by g4s8** [comment]
@olenagerasimova here's full HTTP deploy requests and responses I've collected while deploing test package to Artifactory:
<details>
<summary>HTTP requests</summary>

Format: some requests may include `100` response status before the request, actually it's sent between request head and body because of capture limitations, e.g.
```http
GET /foo HTTP/1.1
Header: value

HTTP/1.1 100 Continue

some-request-body
```
packet is shows as:
```http
HTTP/1.1 100 Continue
GET /foo HTTP/1.1
Header: value

some-request-body
```

```http
HTTP/1.1 100 Continue

PUT /artifactory/libs-release/com/test/logger/1.1/logger-1.1.jar HTTP/1.1
Cache-control: no-cache
Cache-store: no-store
Pragma: no-cache
User-Agent: Apache-Maven/3.6.2 (Java 13.0.1; Linux 5.4.38-gentoo-x86_64)
Content-Length: 2460
Host: localhost:8081
Connection: Keep-Alive
Expect: 100-continue
Accept-Encoding: gzip,deflate
Authorization: Basic YWRtaW46cXdlcnR5MTIz

PK..
.....'..Q;...U...\.......META-INF/MANIFEST.MF.M..LK-..
K-*....R0.3..r.JM,IM.u..Rp,HL.HU.M,K.S0.3.3..r*..).....[@.).^).@..z.z..\.\.PK..;...U...\...PK..
.....'..Q............	...META-INF/PK..
.....'..Q................com/PK..
.....'..Q............	...com/test/PK..
.....'..Q................META-INF/maven/PK..
.....'..Q................META-INF/maven/com.test/PK..
.....'..Q................META-INF/maven/com.test/logger/PK..
.....'..QX.jN............com/test/Log.classmRmO.P.~...[.6..Q.a:]7.!..#|!.S.$%#....\G........Q $...Q.s..e.MO.i..s...n...l.6.AV.f".<..)...!.......$C~..}...mZ."
(.0L..d0.(..H..D.BHN.....2..A..2*:fM<A....=." j4"..[........<..0GVG...j#.@.a`.Ou,.x.....N.r.P.Fq,B...kZ.$..d.L,c.A.....1	....*.t.M..k...>iT_U..a.....39..5..4.)t....D..`..C.7
.">...P#.<>...W...&O....<\.....i.AB.2.h.{b.W..!..56h..."..VMY......P.K..o...Z..........R,A...
.T-..a....Re...4?!...?Q...W.....?.....~Aw{..{....^F..b
..>.;.:..}.....PK..X.jN........PK..
.....'..Q7.w.T...X... ...com/test/com-test-logger-jar.txts*..IQ.+.MJ-.R.+...
(..JM.Q.K-*....R0.3.r..sI,I.R0202.5..50	14..!S....t.....b+.s3c..PK..7.w.T...X...PK..
........Q.8.'K.......&...META-INF/maven/com.test/logger/pom.xml...j.0...y
.{.r.C0.....
....[G....;}.Jv...@o.O.;.J.:s.*dW%..\.B....-j.-.NH.k..y..B	].........t.^.@i....klT|%..\W.\dY.(|.w..A.........vs..J..^...)S.|C......`...r.:.c,.g....+..wAX..np.$......F1..A..L.dM...........03..H.4......'o.n.3wi..\L2.$.'.\.8./..8Y..-.t.D.....%=..k.....Ck...}
 "Q....3..x5.......Pl.6.I.Hq.+...G.s.5.......2..r.
PK...8.'K.......PK..
.....'..Q....O...N...-...META-INF/maven/com.test/logger/pom.propertiesSv.JM,IMQH.Tp,HL.HU.M,K.S0.3.3.J/./-.L.M...+I-..J,*.LKL.........q....g......r..PK......O...N...PK..........'..Q;...U...\.....................META-INF/MANIFEST.MFPK....
.....'..Q............	............A....META-INF/PK....
.....'..Q.........................A....com/PK....
.....'..Q............	............A....com/test/PK....
.....'..Q.........................A....META-INF/maven/PK....
.....'..Q.........................A4...META-INF/maven/com.test/PK....
.....'..Q.........................Aj...META-INF/maven/com.test/logger/PK..........'..QX.jN..........................com/test/Log.classPK..........'..Q7.w.T...X... .................com/test/com-test-logger-jar.txtPK.............Q.8.'K.......&.............N...META-INF/maven/com.test/logger/pom.xmlPK..........'..Q....O...N...-.................META-INF/maven/com.test/logger/pom.propertiesPK....................HTTP/1.1 201 Created
Server: Artifactory/7.4.3 70403900
X-Artifactory-Id: 9e472fba47db7026:45958bf6:173b9abc4ec:-8000
X-Artifactory-Node-Id: 5f090cdcb75c
Location: http://localhost:8081/artifactory/libs-release-local/com/test/logger/1.1/logger-1.1.jar
Content-Type: application/vnd.org.jfrog.artifactory.storage.ItemCreated+json;charset=ISO-8859-1
Transfer-Encoding: chunked
Date: Tue, 04 Aug 2020 14:17:15 GMT

{
  "repo" : "libs-release-local",
  "path" : "/com/test/logger/1.1/logger-1.1.jar",
  "created" : "2020-08-04T13:43:21.413Z",
  "createdBy" : "admin",
  "downloadUri" : "http://localhost:8081/artifactory/libs-release-local/com/test/logger/1.1/logger-1.1.jar",
  "mimeType" : "application/java-archive",
  "size" : "2460",
  "checksums" : {
    "sha1" : "d467b6323d0f9e1fd02ce07e5993d15f2975237c",
    "md5" : "86977953d0a64d940d47462e4456729d",
    "sha256" : "74d7f71379f432b63135f12180305fe8fc52fe7841df2009d1c03730798fe238"
  },
  "originalChecksums" : {
    "sha256" : "74d7f71379f432b63135f12180305fe8fc52fe7841df2009d1c03730798fe238"
  },
  "uri" : "http://localhost:8081/artifactory/libs-release-local/com/test/logger/1.1/logger-1.1.jar"
}HTTP/1.1 100 Continue

PUT /artifactory/libs-release/com/test/logger/1.1/logger-1.1.jar.sha1 HTTP/1.1
Cache-control: no-cache
Cache-store: no-store
Pragma: no-cache
User-Agent: Apache-Maven/3.6.2 (Java 13.0.1; Linux 5.4.38-gentoo-x86_64)
Content-Length: 40
Host: localhost:8081
Connection: Keep-Alive
Expect: 100-continue
Accept-Encoding: gzip,deflate
Authorization: Basic YWRtaW46cXdlcnR5MTIz

d467b6323d0f9e1fd02ce07e5993d15f2975237cHTTP/1.1 201 Created
Server: Artifactory/7.4.3 70403900
X-Artifactory-Id: 9e472fba47db7026:45958bf6:173b9abc4ec:-8000
X-Artifactory-Node-Id: 5f090cdcb75c
Location: http://localhost:8081/artifactory/libs-release-local/com/test/logger/1.1/logger-1.1.jar
Transfer-Encoding: chunked
Date: Tue, 04 Aug 2020 14:17:15 GMT

HTTP/1.1 100 Continue

PUT /artifactory/libs-release/com/test/logger/1.1/logger-1.1.jar.md5 HTTP/1.1
Cache-control: no-cache
Cache-store: no-store
Pragma: no-cache
User-Agent: Apache-Maven/3.6.2 (Java 13.0.1; Linux 5.4.38-gentoo-x86_64)
Content-Length: 32
Host: localhost:8081
Connection: Keep-Alive
Expect: 100-continue
Accept-Encoding: gzip,deflate
Authorization: Basic YWRtaW46cXdlcnR5MTIz

86977953d0a64d940d47462e4456729dHTTP/1.1 201 Created
Server: Artifactory/7.4.3 70403900
X-Artifactory-Id: 9e472fba47db7026:45958bf6:173b9abc4ec:-8000
X-Artifactory-Node-Id: 5f090cdcb75c
Location: http://localhost:8081/artifactory/libs-release-local/com/test/logger/1.1/logger-1.1.jar
Transfer-Encoding: chunked
Date: Tue, 04 Aug 2020 14:17:15 GMT

HTTP/1.1 100 Continue

PUT /artifactory/libs-release/com/test/logger/1.1/logger-1.1.pom HTTP/1.1
Cache-control: no-cache
Cache-store: no-store
Pragma: no-cache
User-Agent: Apache-Maven/3.6.2 (Java 13.0.1; Linux 5.4.38-gentoo-x86_64)
Content-Length: 732
Host: localhost:8081
Connection: Keep-Alive
Expect: 100-continue
Accept-Encoding: gzip,deflate
Authorization: Basic YWRtaW46cXdlcnR5MTIz

<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
  xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
  <modelVersion>4.0.0</modelVersion>
  <parent>
    <groupId>com.artipie</groupId>
    <artifactId>ppom</artifactId>
    <version>0.3.3</version>
  </parent>
  <groupId>com.test</groupId>
  <artifactId>logger</artifactId>
  <packaging>jar</packaging>
  <version>1.1</version>
  <name>logger</name>
  <url>http://maven.apache.org</url>
  <distributionManagement>
    <repository>
      <id>artifactory</id>
      <url>http://localhost:8081/artifactory/libs-release/</url>
    </repository>
  </distributionManagement>
</project>
HTTP/1.1 201 Created
Server: Artifactory/7.4.3 70403900
X-Artifactory-Id: 9e472fba47db7026:45958bf6:173b9abc4ec:-8000
X-Artifactory-Node-Id: 5f090cdcb75c
Location: http://localhost:8081/artifactory/libs-release-local/com/test/logger/1.1/logger-1.1.pom
Content-Type: application/vnd.org.jfrog.artifactory.storage.ItemCreated+json;charset=ISO-8859-1
Transfer-Encoding: chunked
Date: Tue, 04 Aug 2020 14:17:15 GMT

{
  "repo" : "libs-release-local",
  "path" : "/com/test/logger/1.1/logger-1.1.pom",
  "created" : "2020-08-04T13:43:21.837Z",
  "createdBy" : "admin",
  "downloadUri" : "http://localhost:8081/artifactory/libs-release-local/com/test/logger/1.1/logger-1.1.pom",
  "mimeType" : "application/x-maven-pom+xml",
  "size" : "732",
  "checksums" : {
    "sha1" : "f39d8963c61f7fa39dd91c7dfa824c3cde12e922",
    "md5" : "37b7cbc3829ddca579de5c11dad18e37",
    "sha256" : "0e6de0ef6895a016f0fb54b948165cf30ea7daac2e6685c32429786be151baff"
  },
  "originalChecksums" : {
    "sha256" : "0e6de0ef6895a016f0fb54b948165cf30ea7daac2e6685c32429786be151baff"
  },
  "uri" : "http://localhost:8081/artifactory/libs-release-local/com/test/logger/1.1/logger-1.1.pom"
}HTTP/1.1 100 Continue

PUT /artifactory/libs-release/com/test/logger/1.1/logger-1.1.pom.sha1 HTTP/1.1
Cache-control: no-cache
Cache-store: no-store
Pragma: no-cache
User-Agent: Apache-Maven/3.6.2 (Java 13.0.1; Linux 5.4.38-gentoo-x86_64)
Content-Length: 40
Host: localhost:8081
Connection: Keep-Alive
Expect: 100-continue
Accept-Encoding: gzip,deflate
Authorization: Basic YWRtaW46cXdlcnR5MTIz

f39d8963c61f7fa39dd91c7dfa824c3cde12e922HTTP/1.1 201 Created
Server: Artifactory/7.4.3 70403900
X-Artifactory-Id: 9e472fba47db7026:45958bf6:173b9abc4ec:-8000
X-Artifactory-Node-Id: 5f090cdcb75c
Location: http://localhost:8081/artifactory/libs-release-local/com/test/logger/1.1/logger-1.1.pom
Transfer-Encoding: chunked
Date: Tue, 04 Aug 2020 14:17:15 GMT

HTTP/1.1 100 Continue

PUT /artifactory/libs-release/com/test/logger/1.1/logger-1.1.pom.md5 HTTP/1.1
Cache-control: no-cache
Cache-store: no-store
Pragma: no-cache
User-Agent: Apache-Maven/3.6.2 (Java 13.0.1; Linux 5.4.38-gentoo-x86_64)
Content-Length: 32
Host: localhost:8081
Connection: Keep-Alive
Expect: 100-continue
Accept-Encoding: gzip,deflate
Authorization: Basic YWRtaW46cXdlcnR5MTIz

37b7cbc3829ddca579de5c11dad18e37HTTP/1.1 201 Created
Server: Artifactory/7.4.3 70403900
X-Artifactory-Id: 9e472fba47db7026:45958bf6:173b9abc4ec:-8000
X-Artifactory-Node-Id: 5f090cdcb75c
Location: http://localhost:8081/artifactory/libs-release-local/com/test/logger/1.1/logger-1.1.pom
Transfer-Encoding: chunked
Date: Tue, 04 Aug 2020 14:17:15 GMT

GET /artifactory/libs-release/com/test/logger/maven-metadata.xml HTTP/1.1
Cache-control: no-cache
Cache-store: no-store
Pragma: no-cache
User-Agent: Apache-Maven/3.6.2 (Java 13.0.1; Linux 5.4.38-gentoo-x86_64)
Host: localhost:8081
Connection: Keep-Alive
Accept-Encoding: gzip,deflate
Authorization: Basic YWRtaW46cXdlcnR5MTIz

HTTP/1.1 200 OK
Server: Artifactory/7.4.3 70403900
X-Artifactory-Id: 9e472fba47db7026:45958bf6:173b9abc4ec:-8000
X-Artifactory-Node-Id: 5f090cdcb75c
Last-Modified: Tue, 04 Aug 2020 14:17:15 GMT
ETag: df8ab47f1a2f16c2cf653715da35c266be82d2b5
X-Checksum-Sha1: df8ab47f1a2f16c2cf653715da35c266be82d2b5
X-Checksum-Sha256: 54d65960378d8d6dd10cddd7fc1ae32e36ab75772d67e9490a56923b38a66086
X-Checksum-Md5: 5e769b87a99a0463f399e767174cea41
Accept-Ranges: bytes
X-Artifactory-Filename: maven-metadata.xml
Content-Disposition: attachment; filename="maven-metadata.xml"; filename*=UTF-8''maven-metadata.xml
Cache-Control: no-store
Content-Type: application/xml
Content-Length: 361
Date: Tue, 04 Aug 2020 14:17:15 GMT

<?xml version="1.0" encoding="UTF-8"?>
<metadata modelVersion="1.1.0">
  <groupId>com.test</groupId>
  <artifactId>logger</artifactId>
  <version>1.1</version>
  <versioning>
    <latest>1.1</latest>
    <release>1.1</release>
    <versions>
      <version>1.1</version>
    </versions>
    <lastUpdated>20200804141715</lastUpdated>
  </versioning>
</metadata>
GET /artifactory/libs-release/com/test/logger/maven-metadata.xml.sha1 HTTP/1.1
Cache-control: no-cache
Cache-store: no-store
Pragma: no-cache
User-Agent: Apache-Maven/3.6.2 (Java 13.0.1; Linux 5.4.38-gentoo-x86_64)
Host: localhost:8081
Connection: Keep-Alive
Accept-Encoding: gzip,deflate
Authorization: Basic YWRtaW46cXdlcnR5MTIz

HTTP/1.1 200 OK
Server: Artifactory/7.4.3 70403900
X-Artifactory-Id: 9e472fba47db7026:45958bf6:173b9abc4ec:-8000
X-Artifactory-Node-Id: 5f090cdcb75c
Last-Modified: Tue, 04 Aug 2020 14:17:16 GMT
Cache-Control: no-store
Content-Type: application/x-checksum
Content-Length: 40
Date: Tue, 04 Aug 2020 14:17:15 GMT

df8ab47f1a2f16c2cf653715da35c266be82d2b5HTTP/1.1 100 Continue

PUT /artifactory/libs-release/com/test/logger/maven-metadata.xml HTTP/1.1
Cache-control: no-cache
Cache-store: no-store
Pragma: no-cache
User-Agent: Apache-Maven/3.6.2 (Java 13.0.1; Linux 5.4.38-gentoo-x86_64)
Content-Length: 315
Host: localhost:8081
Connection: Keep-Alive
Expect: 100-continue
Accept-Encoding: gzip,deflate
Authorization: Basic YWRtaW46cXdlcnR5MTIz

<?xml version="1.0" encoding="UTF-8"?>
<metadata>
  <groupId>com.test</groupId>
  <artifactId>logger</artifactId>
  <versioning>
    <latest>1.1</latest>
    <release>1.1</release>
    <versions>
      <version>1.1</version>
    </versions>
    <lastUpdated>20200804141716</lastUpdated>
  </versioning>
</metadata>
HTTP/1.1 201 Created
Server: Artifactory/7.4.3 70403900
X-Artifactory-Id: 9e472fba47db7026:45958bf6:173b9abc4ec:-8000
X-Artifactory-Node-Id: 5f090cdcb75c
Location: http://localhost:8081/artifactory/libs-release-local/com/test/logger/maven-metadata.xml
Content-Type: application/vnd.org.jfrog.artifactory.storage.ItemCreated+json;charset=ISO-8859-1
Transfer-Encoding: chunked
Date: Tue, 04 Aug 2020 14:17:15 GMT

{
  "repo" : "libs-release-local",
  "path" : "/com/test/logger/maven-metadata.xml",
  "created" : "2020-08-04T13:43:21.960Z",
  "createdBy" : "admin",
  "downloadUri" : "http://localhost:8081/artifactory/libs-release-local/com/test/logger/maven-metadata.xml",
  "mimeType" : "application/xml",
  "size" : "315",
  "checksums" : {
    "sha1" : "fe6ffac55e6be66a349b868e1b233f9f0ea01e16",
    "md5" : "413d164d5f5f592f9fed99e5c73d8619",
    "sha256" : "13081d5ae4b790d2ce610e66ef80b5a5d4272325594b05d1aca09c0a95f55ca6"
  },
  "originalChecksums" : {
    "sha256" : "13081d5ae4b790d2ce610e66ef80b5a5d4272325594b05d1aca09c0a95f55ca6"
  },
  "uri" : "http://localhost:8081/artifactory/libs-release-local/com/test/logger/maven-metadata.xml"
}HTTP/1.1 100 Continue

PUT /artifactory/libs-release/com/test/logger/maven-metadata.xml.sha1 HTTP/1.1
Cache-control: no-cache
Cache-store: no-store
Pragma: no-cache
User-Agent: Apache-Maven/3.6.2 (Java 13.0.1; Linux 5.4.38-gentoo-x86_64)
Content-Length: 40
Host: localhost:8081
Connection: Keep-Alive
Expect: 100-continue
Accept-Encoding: gzip,deflate
Authorization: Basic YWRtaW46cXdlcnR5MTIz

fe6ffac55e6be66a349b868e1b233f9f0ea01e16HTTP/1.1 200 OK
Server: Artifactory/7.4.3 70403900
X-Artifactory-Id: 9e472fba47db7026:45958bf6:173b9abc4ec:-8000
X-Artifactory-Node-Id: 5f090cdcb75c
Transfer-Encoding: chunked
Date: Tue, 04 Aug 2020 14:17:15 GMT

HTTP/1.1 100 Continue

PUT /artifactory/libs-release/com/test/logger/maven-metadata.xml.md5 HTTP/1.1
Cache-control: no-cache
Cache-store: no-store
Pragma: no-cache
User-Agent: Apache-Maven/3.6.2 (Java 13.0.1; Linux 5.4.38-gentoo-x86_64)
Content-Length: 32
Host: localhost:8081
Connection: Keep-Alive
Expect: 100-continue
Accept-Encoding: gzip,deflate
Authorization: Basic YWRtaW46cXdlcnR5MTIz

413d164d5f5f592f9fed99e5c73d8619HTTP/1.1 200 OK
Server: Artifactory/7.4.3 70403900
X-Artifactory-Id: 9e472fba47db7026:45958bf6:173b9abc4ec:-8000
X-Artifactory-Node-Id: 5f090cdcb75c
Transfer-Encoding: chunked
Date: Tue, 04 Aug 2020 14:17:16 GMT
```
</details>

**Comment by 0pdd** [comment]
@g4s8 2 puzzles [#124](https://github.com/artipie/maven-adapter/issues/124), [#125](https://github.com/artipie/maven-adapter/issues/125) are still not solved.

**Comment by 0pdd** [comment]
@g4s8 the puzzle [#125](https://github.com/artipie/maven-adapter/issues/125) is still not solved; solved: [#124](https://github.com/artipie/maven-adapter/issues/124).

**Comment by 0pdd** [comment]
@g4s8 2 puzzles [#131](https://github.com/artipie/maven-adapter/issues/131), [#132](https://github.com/artipie/maven-adapter/issues/132) are still not solved; solved: [#124](https://github.com/artipie/maven-adapter/issues/124), [#125](https://github.com/artipie/maven-adapter/issues/125).

**Comment by 0pdd** [comment]
@g4s8 the puzzle [#131](https://github.com/artipie/maven-adapter/issues/131) is still not solved; solved: [#124](https://github.com/artipie/maven-adapter/issues/124), [#125](https://github.com/artipie/maven-adapter/issues/125), [#132](https://github.com/artipie/maven-adapter/issues/132).

**Comment by olenagerasimova** [comment]
@g4s8 
> check metadata group and id is the same as in repo maven-metadata.xml file, check metadata versions are correct

question about checking uploaded metadata versions: now we simply generate new versions list by listing all repository items by package key. How are you suggesting to validate versions in `maven-metadata.xml`?

**Comment by g4s8** [comment]
@olenagerasimova I think we can validate version by reading uploaded client `maven-metadata.xml` and checking it contains artifact version from upload path, e.g. if client uploads `com/foo/bar/1.0/bar-1.0.jar` but `maven-metadata.xml` has `0.1` version we can treat such upload as invalid.

**Comment by olenagerasimova** [comment]
@g4s8 as files are uploaded one by one independently, I read version from `maven-metadata.xml` to start upload validation process (check AstoValidUpload#validateChecksums), so we will validate upload by version specified in `maven-metadata.xml`. The only thing we can do here - treat upload invalid if no artifacts were found by specified version.

**Comment by g4s8** [comment]
@olenagerasimova yes, it makes sense. In such case can you please propose some validation flow for new artifact deploy?

**Comment by olenagerasimova** [comment]
@g4s8 sure, I'll submit PR, describe it there and we can discuss any details.

**Comment by 0pdd** [comment]
@g4s8 the puzzle [#143](https://github.com/artipie/maven-adapter/issues/143) is still not solved; solved: [#124](https://github.com/artipie/maven-adapter/issues/124), [#125](https://github.com/artipie/maven-adapter/issues/125), [#131](https://github.com/artipie/maven-adapter/issues/131), [#132](https://github.com/artipie/maven-adapter/issues/132).

**Comment by 0pdd** [comment]
@g4s8 all 5 puzzles are solved here: [#124](https://github.com/artipie/maven-adapter/issues/124), [#125](https://github.com/artipie/maven-adapter/issues/125), [#131](https://github.com/artipie/maven-adapter/issues/131), [#132](https://github.com/artipie/maven-adapter/issues/132), [#143](https://github.com/artipie/maven-adapter/issues/143).
